### PR TITLE
feat: implement image cache configuration

### LINKED ISSUE #9767: image cache: volume config, mounting, feature, image cache config

### PR DESCRIPTION
**Pull request by smira**
Implement a feature flag, a resource which controls the flow.

This controls the volume configuration, mounting, etc.

Fixes #9767

### CODE REVIEW

**Comment by smira** [comment]
Totally WIP at the moment.

File: internal/app/machined/pkg/controllers/cri/image_cache_config.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+		if err != nil && !state.IsNotFoundError(err) {
+			if state.IsNotFoundError(err) {
+				continue
+			}
+
+			return fmt.Errorf("error getting config: %w", err)
+		}
+
+		// image cache is disabled
+		if cfg == nil || cfg.Config().Machine() == nil || !cfg.Config().Machine().Features().ImageCacheEnabled() {
```
Comment (frezbo) [inline internal/app/machined/pkg/controllers/cri/image_cache_config.go:93]: hmm i guess this means we disable on a boot from iso?

File: internal/app/machined/pkg/controllers/block/internal/volumes/format.go [hunk 2]
```diff
@@ -148,6 +158,14 @@ func GrowFilesystem(logger *zap.Logger, volumeContext ManagerContext) error {
 			return fmt.Errorf("error growing XFS: %w", err)
 		}
 
+		return nil
+	case block.FilesystemTypeEXT4:
+		logger.Info("growing ext4 filesystem", zap.String("device", volumeContext.Status.MountLocation))
+
+		if err := makefs.Ext4Resize(volumeContext.Status.MountLocation); err != nil {
```
Comment (frezbo) [inline internal/app/machined/pkg/controllers/block/internal/volumes/format.go:165]: resize2fs works on the blockdevice, not the mount as like xfs

Comment (smira) [reply internal/app/machined/pkg/controllers/block/internal/volumes/format.go:165]: yes, this is how it works here - on the blockdevice (it doesn't do any mounts)

Comment (frezbo) [reply internal/app/machined/pkg/controllers/block/internal/volumes/format.go:165]: ahh okay, i assumed `volumeContext.Status.MountLocation` was the mounted location

Comment (smira) [reply internal/app/machined/pkg/controllers/block/internal/volumes/format.go:165]: it's the mount source, e.g. for encrypted devices it's different from "Location" which is the encrypted blockdevice

**Comment by smira** [comment]
```yaml
# talosctl -n 172.20.0.5 get imagecacheconfig -o yaml 
node: 172.20.0.5
metadata:
    namespace: cri
    type: ImageCacheConfigs.cri.talos.dev
    id: image-cache
    version: 4
    owner: cri.ImageCacheConfigController
    phase: running
    created: 2024-11-21T17:21:14Z
    updated: 2024-11-21T17:21:17Z
spec:
    status: ready
    roots:
        - /system/imagecache/disk
```

**Comment by smira** [comment]
```
$ talosctl -n 172.20.0.5 get mountstatus                  
NODE         NAMESPACE   TYPE          ID           VERSION   SOURCE           TARGET                    FILESYSTEM TYPE
172.20.0.5   runtime     MountStatus   EPHEMERAL    1         /dev/vda6        /var                      xfs
172.20.0.5   runtime     MountStatus   IMAGECACHE   1         /dev/nvme0n1p1   /system/imagecache/disk   ext4
172.20.0.5   runtime     MountStatus   STATE        1         /dev/vda5        /system/state             xfs
```

**Comment by frezbo** [comment]
```yaml
machine:
  features:
    imageCache: true
```

seems this patch needs a reboot, I guess that was not intended :thinking:

**Comment by smira** [comment]
> ```yaml
> machine:
>   features:
>     imageCache: true
> ```
> 
> seems this patch needs a reboot, I guess that was not intended 🤔

ah, true, I'll get it fixed

**Pull request merged**